File: mac/src/Honeycomb/Models/HoneycombCard.swift
import Foundation

public struct HoneycombCardData: Codable, Identifiable, Equatable {
    public let id: Int
    public let name: String
    public let stars: Int
    public let stats: [Int] // 0: Top, 1: Right, 2: Bottom, 3: Left
    public let suit: String

    public static func suitDisplayName(_ code: String) -> String {
        switch code {
        case "S": return "Spades"
        case "H": return "Hearts"
        case "D": return "Diamonds"
        case "C": return "Clubs"
        default: return code
        }
    }
}

public enum CardOwner: String, Codable, Equatable {
    case player
    case opponent
}

public struct HoneycombCard: Codable, Identifiable, Equatable {
    public let id: String
    public let data: HoneycombCardData
    public var owner: CardOwner
    public let originalOwner: CardOwner
    public var modifier: Int = 0

    // originalOwner defaults to owner (the normal case), but the "Swap" rule needs to
    // construct a card whose current holder (owner, for battle purposes) differs from
    // its true owner (originalOwner) — it's played by whoever it was swapped to, but
    // reverts to its rightful owner for the "keep what you hold" win-unlock (and stays
    // stealable by that rightful owner if they don't recapture it themselves).
    public init(data: HoneycombCardData, owner: CardOwner, originalOwner: CardOwner? = nil, id: String = UUID().uuidString) {
        self.data = data
        self.owner = owner
        self.originalOwner = originalOwner ?? owner
        self.id = id
    }

    public func stat(at index: Int) -> Int {
        let val = data.stats[index] + modifier
        // Matches FFXIV Triple Triad: stats are clamped to the 1-10 (A) range in all
        // capture math, so an Ascension-boosted 8 behaves exactly as a 10, not an 11.
        return min(10, max(1, val))
    }
}

File: windows/src/SoliBee.Core/Models/HoneycombCard.cs
using System;

namespace SoliBee.Core.Models;

public class HoneycombCardData
{
    public int Id { get; init; }
    public string Name { get; init; } = "";
    public int Stars { get; init; }
    public int[] Stats { get; init; } = new int[4];
    public string Suit { get; init; } = "";

    public static string SuitDisplayName(string code) => code switch
    {
        "S" => "Spades",
        "H" => "Hearts",
        "D" => "Diamonds",
        "C" => "Clubs",
        _ => code
    };
}

public class HoneycombCard
{
    public HoneycombCardData Data { get; }
    public int Modifier { get; set; }
    public int OriginalOwner { get; set; }
    public int Owner { get; set; }
    public Guid UniqueInstanceId { get; set; } = Guid.NewGuid();
    public bool IsFaceDown { get; set; } = false;

    // Bomb Shelter: turns left until this face-down card flips automatically. Set to 3
    // when placed face-down, decremented once per subsequent play (either side), and
    // cleared (null) once revealed — null means "not a timed Bomb Shelter card".
    public int? BombShelterTurnsRemaining { get; set; }

    public HoneycombCard(HoneycombCardData data, int owner)
    {
        Data = data;
        Owner = owner;
        OriginalOwner = owner;
    }

    public int Stat(int index)
    {
        int val = Data.Stats[index] + Modifier;
        return Math.Min(10, Math.Max(1, val));
    }

    public HoneycombCard Clone()
    {
        return new HoneycombCard(Data, Owner)
        {
            Modifier = Modifier,
            OriginalOwner = OriginalOwner,
            UniqueInstanceId = UniqueInstanceId,
            IsFaceDown = IsFaceDown,
            BombShelterTurnsRemaining = BombShelterTurnsRemaining
        };
    }
}
